src/core/algorithms/od/fastod/range_based_stripped_partition.cpp

```diff
@@ -0,0 +1,261 @@
+
+    indexes_ = std::move(new_indexes);
+    begins_ = std::move(new_begins);
+}
+
+bool RangeBasedStrippedPartition::Split(short right) const {
+    for (size_t begin_pointer = 0; begin_pointer < begins_.size() - 1; ++begin_pointer) {
+        size_t group_begin = begins_[begin_pointer];
+        size_t group_end = begins_[begin_pointer + 1];
+
+        int group_value = data_.GetValue(indexes_[group_begin].first, right);
+
+        for (size_t i = group_begin; i < group_end; ++i) {
+            DataFrame::range_t range = indexes_[i];
+
+            for (size_t j = range.first; j <= range.second; ++j) {
+                if (data_.GetValue(j, right) != group_value) {
+                    return true;
+                }
+            }
+        }
+    }
+
+    return false;
+}
+
+bool RangeBasedStrippedPartition::Swap(short left, short right, bool ascending) const {
+    for (size_t begin_pointer = 0; begin_pointer < begins_.size() - 1; begin_pointer++) {
+        size_t group_begin = begins_[begin_pointer];
+        size_t group_end = begins_[begin_pointer + 1];
+
+        std::vector<std::pair<int, int>> values;
+        // values.reserve(...)?
+
+        for (size_t i = group_begin; i < group_end; ++i) {
+            DataFrame::range_t range = indexes_[i];
+
+            for (size_t j = range.first; j <= range.second; ++j) {
+                values.push_back({data_.GetValue(j, left), data_.GetValue(j, right)});
+            }
+        }
+
+        if (ascending) {
+            std::sort(values.begin(), values.end(),
+                      [](auto const& p1, auto const& p2) { return p1.first < p2.first; });
+        } else {
+            std::sort(values.begin(), values.end(),
+                      [](auto const& p1, auto const& p2) { return p2.first < p1.first; });
+        }
+
+        size_t prev_group_max_index = 0;
+        size_t current_group_max_index = 0;
+        bool is_first_group = true;
+
+        for (size_t i = 0; i < values.size(); i++) {
+            auto const& first = values[i].first;
+            auto const& second = values[i].second;
+
+            if (i != 0 && values[i - 1].first != first) {
+                is_first_group = false;
+                prev_group_max_index = current_group_max_index;
+                current_group_max_index = i;
+            } else if (values[current_group_max_index].second <= second) {
+                current_group_max_index = i;
+            }
+
+            if (!is_first_group && values[prev_group_max_index].second > second) {
+                return true;
+            }
+        }
+    }
+
+    return false;
+}
+
+std::vector<DataFrame::value_indexes_t> RangeBasedStrippedPartition::IntersectWithAttribute(
+        algos::fastod::AttributeSet::size_type attribute, size_t group_start, size_t group_end) {
+    std::vector<DataFrame::value_indexes_t> result;
+    std::vector<DataFrame::value_indexes_t> const& attr_ranges =
+            data_.GetDataRanges().at(attribute);
```

Comment:
Why not `operator[]`?

Reply:
No specific reason. Now we use `operator[]`.